### ISSUE TITLE
Need some help to fully re-implement the FaceNet paper.

### ISSUE THREAD

**Issue by myme5261314** [davidsandberg/facenet#14]
Hey David,
   Thanks for your code that combines with the `openface` code to let me investigate.
   I've forked your facenet and been working on my own for about 2-3 weeks. I've followed the torch code from `openface` and seems re-implemented their strategy of training. Especially the one in the image below.
![sad](http://bamos.github.io/data/2016-01-19/optimization-after.png)

I see your code have some bugs and weird strategy for me.
1. the batch norm code you borrow from the stackoverflow is wrong, refer to @jrocks' answer on the question for detail. [link](http://stackoverflow.com/questions/33949786/how-could-i-use-batch-normalization-in-tensorflow)
2. I see you want to use 45\*40 images to generate triplets, but limited by GPU memory, you doesn't follow openface triplets selection strategy but use every first image of certain people to be the `anchor`, this seems not very reasonable right? I mean `anchor` doesn't mean not changed.
3. This is the most serious problem I'm facing. You split selecting triplets by mini-batch (say 90), and then backward one mini-batch a time. What I have done in my code is that expanded triplets and then accumulate gradient coefficients input just like the `openface` and then calculate the gradients for all variables by feeding 20 mini-batches (20\*90=45\*40), then accumulate these gradients, then apply them once. This procedure is extended from `openface` since they use 15\*20 images, and their torch SDK seems have no API to get gradients. But for now my code isn't working, means the training doesn't seems to achieve at least 80% accuracy on LFW, I don't know why is that, could you please help me out?

You may check the code from [my fork optimize branch](https://github.com/myme5261314/facenet/tree/optimize). Sorry that I've heavily modified the file structure and python coding style.

Please contact me if you have any idea. I think we could work together to use both 45\*40 batch size and non-compact triplets.

**Comment by RomainSabathe** [comment]
Hi myme5261314, all the questions you asked are really relevant. Did you get any insights or improvements?

**Comment by myme5261314** [comment]
@RomainSabathe Because of some personal reason, I had stopped doing about facenet stuff since a few months ago. Before I quit, I can get no more accuracy than openface had achieved that time which is around 97% on LFW. Seems no significant improvement by re-implement the `openface` approach through `facenet` and use 1800 batch size. Hope this can give you the info you want.

**Comment by RomainSabathe** [comment]
@myme5261314 that's helpful, thank you very much. :)
